cmp/grouping/GroupingChooserModel.js

```diff
@@ -0,0 +1,285 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+
+import {HoistModel, managed, PersistenceProvider, XH} from '@xh/hoist/core';
+import {action, computed, observable} from '@xh/hoist/mobx';
+import {genDisplayName} from '@xh/hoist/data';
+import {throwIf} from '@xh/hoist/utils/js';
+import {createObservableRef} from '@xh/hoist/utils/react';
+import {cloneDeep, difference, isArray, isEmpty, isEqual, isString, keys, sortBy} from 'lodash';
+
+@HoistModel
+export class GroupingChooserModel {
+
+    /** @member {string[]} - names of dimensions selected for grouping. */
+    @observable.ref value;
+    /** @member {string[][]} - array of dim-name value arrays */
+    @observable.ref favorites = [];
+
+    /** @member {number} */
+    maxDepth;
+    /** @member {Object<string, DimensionSpec>} */
+    dimensions;
+    /** @member {string[]} */
+    dimensionNames;
+    /** @member {PersistenceProvider} */
+    @managed provider = null;
+    persistValue = false;
+    persistFavorites = false;
+
+    // Implementation fields for Control
+    @observable.ref pendingValue = [];
+    @observable editorIsOpen = false;
+    @observable favoritesIsOpen = false;
+    @observable showAddControl = false;
+
+    popoverRef = createObservableRef();
+
+    @computed
+    get availableDims() {
+        return difference(this.dimensionNames, this.pendingValue);
+    }
+
+    @computed
+    get atMaxDepth() {
+        return this.pendingValue.length === Math.min(this.maxDepth, this.dimensionNames.length);
+    }
+
+    @computed
+    get isValid() {
+        return this.validateValue(this.pendingValue);
+    }
+
+    @computed
+    get addDisabledMsg() {
+        if (isEmpty(this.availableDims)) return 'All dimensions added';
+        if (this.atMaxDepth) return 'Grouping at max depth';
+        return null;
+    }
+
+    /**
+     * @param c - GroupingChooserModel configuration.
+     * @param {(DimensionSpec[]|CubeField[]|string[])} c.dimensions - dimensions available for
+     *     selection. When using GroupingChooser to create Cube queries, it is recommended to pass
+     *     the `dimensions` from the related cube (or a filtered subset thereof).
```

Comment:
Thoughts on renaming this `fields` or `fieldSpecs`?  I am *not* sure about this, but wanted to raise the idea in case  people were sympathetic to it and we wanted to take this opportunity to standardize a bit more.